### PR TITLE
Parametric shapes refactoring

### LINKED ISSUE #346: Adding extrude both as an options
This is a nice and easy issue ideal for a first contribution 

The ExtrudeMixedShape and the ExtrudeSplineShape could do with an optional argument that allow extrusion in one or both directions.

Both directions is currently hard coded in these two classes

These two files that will need changing

https://github.com/ukaea/paramak/blob/develop/paramak/parametric_shapes/extruded_spline_shape.py
https://github.com/ukaea/paramak/blob/develop/paramak/parametric_shapes/extruded_mixed_shape.py

There are two other extrude shapes that already have this feature so might be useful (ExtrudeStraightShape and ExtrudeCirlceShape)

Here are the relevant lines from ExtrudeStraightShape that can be copied

https://github.com/ukaea/paramak/blob/12b75aa2a2bee7c589b0e3ff7fce2fd9dc0d77e8/paramak/parametric_shapes/extruded_straight_shape.py#L21
https://github.com/ukaea/paramak/blob/12b75aa2a2bee7c589b0e3ff7fce2fd9dc0d77e8/paramak/parametric_shapes/extruded_straight_shape.py#L42
https://github.com/ukaea/paramak/blob/12b75aa2a2bee7c589b0e3ff7fce2fd9dc0d77e8/paramak/parametric_shapes/extruded_straight_shape.py#L82
https://github.com/ukaea/paramak/blob/12b75aa2a2bee7c589b0e3ff7fce2fd9dc0d77e8/paramak/parametric_shapes/extruded_straight_shape.py#L105

### PR DESCRIPTION
**Pull request by RemDelaporteMathurin**
## Proposed changes

This PR proposes a refactoring of the parametric shapes.
Instead of having parametric shapes inherit from shape, the non-mixed shapes (eg. `ExtrudeStraightShape`, `RotateSplineShape`) now inherit from the mixed shape (eg, `ExtrudeMixedShape`, `RotateMixedShape`).

This allows to avoid code repetition (`create_solid` method).

`RotateCircleShape` and `ExtrudeCircleShape` aren't impacted by this refactoring since these classes aren't used in the same way.

Others:
- Fixes #346 

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Pep8 applied
- [x] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

ToDo: 
- points cannot be set as a required argument (ie. without a default value) cause the components don't find points before initialisation --> if `points` is defaulted to `None` or `[]`, `points=[(*p, "spline") for p in points]` won't work
- adding "spline", "straight" must be done every time `points` is modified.

### CODE REVIEW

**Comment by RemDelaporteMathurin** [comment]
@billingsley-john I might need your advice here

**Comment by RemDelaporteMathurin** [comment]
@Shimwell @billingsley-john I think this is ready for review

**Comment by Shimwell** [comment]
Always coming up with nice refactoring methods, will take a peak soon. Thanks for this

File: paramak/parametric_components/tokamak_plasma.py [hunk 1]
```diff
@@ -73,15 +73,15 @@ def __init__(
...
-    # overwrite points setter in Shape (error returned if not overwritten)
-    @property
-    def points(self):
-        self.find_points()
-        return self._points
+    # # overwrite points setter in Shape (error returned if not overwritten)
+    # @property
+    # def points(self):
+    #     self.find_points()
+    #     return self._points
```
Comment (Shimwell) [inline paramak/parametric_components/tokamak_plasma.py:80]: are these still needed
```suggestion
```

File: paramak/parametric_components/tokamak_plasma.py [hunk 2]
```diff
@@ -73,15 +73,15 @@ def __init__(
...
 
-    @points.setter
-    def points(self, value):
-        self._points = value
+    # @points.setter
+    # def points(self, value):
+    #     self._points = value
```
Comment (Shimwell) [inline paramak/parametric_components/tokamak_plasma.py:84]: are these still needed
```suggestion
```

File: paramak/parametric_components/tokamak_plasma.py [hunk 3]
```diff
@@ -183,8 +183,7 @@ def Z(theta):
         points = np.stack((R_points, Z_points), axis=1)
 
         # set self.points
-        self.points = list(points)
-
+        self.points = [(p[0], p[1]) for p in points[:-1]]
```
Comment (Shimwell) [inline paramak/parametric_components/tokamak_plasma.py:186]: any chance you could add a comment to say why all the points are not used

**Comment by Shimwell** [comment]
Lots of great stuff here. I'm wondering if this is compatible with the hash value. We perhaps don't have tests for this but when a parametric shape is made it won't be remade unless the inputs change. This is controled with a hash value, @billingsley-john

**Comment by Shimwell** [comment]
Working progress but trying to draw the new layout https://lucid.app/invitations/accept/7a624a34-7695-4ca4-9544-c3c8887c0eaf using the openmc_workshop account. Looks like it is the way to go, especially now that the sweep methods are close to being merged.

File: paramak/parametric_components/tokamak_plasma.py (hunk 3 above)
Comment (RemDelaporteMathurin) [reply paramak/parametric_components/tokamak_plasma.py:177]: I added a comment. This is related with the PR i opened a while ago : if points[0] == points[-1] and error is raised (even though in the end, points[0] is appended at the end of points...). I still don't think an error should be raised here

**Comment by Shimwell** [comment]
merging into a local branch will open a new PR

**Pull request merged**